goship.go

```diff
@@ -760,23 +791,216 @@ func extractOutputHandler(fn func(http.ResponseWriter, *http.Request, string, st
 	}
 }
 
+type User struct {
+	UserName, UserAvatar string
+	auth                 auth
+}
+
+func getUser(r *http.Request) (User, error) {
+	u := User{}
+	if authentication.authorization != true {
+		u.UserName = *defaultUser
+		u.UserAvatar = *defaultAvatar
+		return u, nil
+	}
+	session, err := store.Get(r, sessionName)
+	if err != nil {
+		return u, errors.New("Error Getting User Session")
+	}
+	if _, ok := session.Values["userName"]; !ok {
+		return u, errors.New("No username")
+	}
+	if _, ok := session.Values["avatarURL"]; !ok {
+		return u, errors.New("No avatar")
+	}
+	// Check return of nil
+	u.UserName = session.Values["userName"].(string)
+	u.UserAvatar = session.Values["avatarURL"].(string)
+	return u, nil
+}
+
+//Used to wrap a view and check for auth
+func checkAuth(fn http.HandlerFunc, a auth) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		_, err := getUser(r)
+		if err != nil {
+			log.Printf("error getting a User %s", err)
+			http.Redirect(w, r, os.Getenv("GITHUB_CALLBACK_URL")+"/auth/github/login", http.StatusMovedPermanently)
+
+			return
+		}
+		fn.ServeHTTP(w, r)
+	})
+}
+
+// remove projects where user has no access
+func cleanProjects(cp []goship.Project, r *http.Request, u User) []goship.Project {
+	if authentication.authorization != true {
+		return cp
+	}
+	cleanProjects := []goship.Project{}
+	for _, p := range cp {
+		a := isCollaborator(p.RepoOwner, p.RepoName, u.UserName)
+		// If user does not have access to the project remove it.
+		if a == true {
+			cleanProjects = append(cleanProjects, p)
+		}
+	}
+	return cleanProjects
+}
+
+func loginHandler(providerName string, auth bool) http.HandlerFunc {
+
+	if auth != true {
+		return func(w http.ResponseWriter, r *http.Request) {}
+	}
+
+	provider, err := gomniauth.Provider(providerName)
+	if err != nil {
+		panic(err)
+	}
+	return func(w http.ResponseWriter, r *http.Request) {
+
+		state := gomniauth.NewState("after", "success")
+
+		authUrl, err := provider.GetBeginAuthURL(state, nil)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		http.Redirect(w, r, authUrl, http.StatusFound)
+	}
+}
+
+func callbackHandler(providerName string, auth bool) http.HandlerFunc {
+
+	if auth != true {
+		return func(w http.ResponseWriter, r *http.Request) {}
+	}
+
+	provider, err := gomniauth.Provider(providerName)
+	if err != nil {
+		panic(err)
```

Comment:
@ramen74 panicking seems like overkill here as well. 500 error with a log would be fine i think